sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp

```diff
@@ -0,0 +1,257 @@
+//===------- matrix-unified.hpp - SYCL matrix extension ----*- C++ -*------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+#include <sycl/ext/oneapi/matrix/matrix-tensorcores.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace matrix {
+
+template <typename T, use Use, size_t Rows, size_t Cols,
+          layout Layout = layout::dynamic, typename Group = sycl::sub_group>
+struct joint_matrix {
+
+#if defined(__SYCL_DEVICE_ONLY__)
+#if defined(__NVPTX__)
+  sycl::ext::oneapi::detail::joint_matrix_cuda<T, Use, Rows, Cols, Layout,
+                                               Group>
+      cuda_impl;
+#endif
+#endif
+
+  joint_matrix() {
+#ifndef __SYCL_DEVICE_ONLY__
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  joint_matrix(Group sg) {
+#ifndef __SYCL_DEVICE_ONLY__
+    std::ignore = sg;
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  inline __SYCL_ALWAYS_INLINE decltype(auto) get_wi_data() {
+#if defined(__SYCL_DEVICE_ONLY__)
+#if defined(__NVPTX__)
+    return wi_data(cuda_impl.wi_marray);
+#else
+//  intel impl TODO
+#endif
+#else
+    // Host version of get_wi_data required by compiler even though it will
+    // never be called because joint_matrix cannot be constructed on host.
+    if constexpr (std::is_same_v<T, precision::tf32>) {
+      marray<float, 1> unused{};
+      return wi_data<float, 1>(unused);
+    } else {
+      marray<T, 1> unused{};
+      return wi_data<T, 1>(unused);
+    }
```

Comment:
No this doesn't work, because the host compilation doesn't know about `__NVPTX__`. In any case I don't think this is necessary. The host code will never be called at runtime, since joint_matrix won't be constructable: so all will need is that it can be compiled: I think you will find that you can use the `wi_data<T,1>` return type for all your test cases no problem: both versions of wi_data have the same interfaces, and support whatever data type T is passed to them.

Reply:
If it does turn out there is some problem with this approach, then it would be easy to solve anyway. I think that the key thing is to decide on the user facing interfaces.

Reply:
>I think you will find that you can use the `wi_data<T,1>` return type for all your test cases no problem: both versions of wi_data have the same interfaces, and support whatever data type T is passed to them.

intel and nv's wi_data has different template list. if we keep the code here, our testcase will compfail on host code:
```
bash-4.4$ clang++ -fsycl -O2 sycl/test/matrix/matrix-int8-test-use.cpp -DSYCL_EXT_ONEAPI_MATRIX_VERSION=2
In file included from sycl/test/matrix/matrix-int8-test-use.cpp:8:
In file included from /export/users/yubing/intelllvm/build/bin/../include/sycl/sycl.hpp:69:
In file included from /export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix.hpp:26:
/export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix-unified.hpp:59:14: error: too few template arguments for class template 'wi_data'
      return wi_data<float, 1>(unused);
             ^
/export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix-intel.hpp:307:7: note: template is declared here
class wi_data {
      ^
In file included from sycl/test/matrix/matrix-int8-test-use.cpp:8:
In file included from /export/users/yubing/intelllvm/build/bin/../include/sycl/sycl.hpp:69:
In file included from /export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix.hpp:26:
/export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix-unified.hpp:62:14: error: too few template arguments for class template 'wi_data'
      return wi_data<T, 1>(unused);
             ^
/export/users/yubing/intelllvm/build/bin/../include/sycl/ext/oneapi/matrix/matrix-intel.hpp:307:7: note: template is declared here
class wi_data {
```
i am ok with the patch except for this "get_wi_data"issue

Reply:
@JackAKirk, since the get_wi_data per the spec should return wi_data<T, Rows, Cols, Use, Layout, Group>(*this); 
isn't this what the host implementation should return?
To make the host compilation pass in the case of CUDA implementation, can you add in tensor core header, in the wi_data class, the if statement: #ifdef __SYCL_DEVICE_ONLY__ //use CUDA definition 
#else //use spec definition
so in unified header, in the lines Bing is pointing to: we use the spec definition.

Does this make sense?

Reply:
Yes, I agree, get_wi_marray (and other things that are not in the unified spec) should move to the cuda:matrix namespace.
As for the compiler hack, why we are needing that in first place? in other functions, we throw a runtime error on the host. But for get_wi_data, we are doing something different.

Reply:
Actually it is really the same for get_wi_data as the other cases: They need a return type to compile. get_wi_data uses return type inference rather, that is the only difference; you can add a runtime error line if you want, but it will do nothing, since the code cannot be called because a joint_matrix can't be instantiated on host. Return type inference was used to allow for the two versions of wi_data.